internal/plugin/builtin/pluginapigen/handle.go

```diff
@@ -0,0 +1,256 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package pluginapigen provides a plugin Handle that generates code used by the
+// plugin system itself.
+//
+// This is made available as a hidden flag: "--generate-plugin-api".
+package pluginapigen
+
+import (
+	"path/filepath"
+	"strings"
+
+	iplugin "github.com/thriftrw/thriftrw-go/internal/plugin"
+	"github.com/thriftrw/thriftrw-go/plugin"
+	"github.com/thriftrw/thriftrw-go/plugin/api"
+)
+
+// Handle is a plugin.Handle that generates code for the plugin system API.
+var Handle iplugin.Handle = handle{}
+
+type handle struct{}
+
+func (handle) Name() string {
+	return "pluginapigen"
+}
+
+func (handle) Close() error {
+	return nil // no-op
+}
+
+func (handle) ServiceGenerator() iplugin.ServiceGenerator {
+	return sgen{}
+}
+
+type sgen struct{}
+
+func (sgen) Handle() iplugin.Handle {
+	return Handle
+}
+
+func (sgen) Generate(req *api.GenerateServiceRequest) (*api.GenerateServiceResponse, error) {
+	files := make(map[string][]byte)
+	for _, serviceID := range req.RootServices {
+		service := req.Services[serviceID]
+		module := req.Modules[service.ModuleID]
+
+		templateData := struct {
+			Service *api.Service
+			Request *api.GenerateServiceRequest
+		}{Service: service, Request: req}
+
+		var (
+			err       error
+			ifaceOpts []plugin.TemplateOption
+		)
+
+		ifacePath := filepath.Join(module.Directory, strings.ToLower(service.Name)+".go")
+		ifaceOpts = append(ifaceOpts, templateOptions...)
+		ifaceOpts = append(ifaceOpts, plugin.GoFileImportPath(module.Package))
+		files[ifacePath], err = plugin.GoFileFromTemplate(
+			ifacePath, interfaceTemplate, templateData, ifaceOpts...)
+		if err != nil {
+			return nil, err
+		}
+
+		clientPath := filepath.Join(service.Directory, "client.go")
+		files[clientPath], err = plugin.GoFileFromTemplate(
+			clientPath, clientTemplate, templateData, templateOptions...)
+		if err != nil {
+			return nil, err
+		}
+
+		handlerPath := filepath.Join(service.Directory, "handler.go")
+		files[handlerPath], err = plugin.GoFileFromTemplate(
+			handlerPath, handlerTemplate, templateData, templateOptions...)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &api.GenerateServiceResponse{Files: files}, nil
+}
+
+// convenience function because "index .Request.Services .Service.ParentID"
+// doesn't work. Index expects a ServiceID, not *ServiceID.
+func getService(req *api.GenerateServiceRequest, id api.ServiceID) *api.Service {
+	return req.Services[id]
+}
+
+var templateOptions = []plugin.TemplateOption{
+	plugin.TemplateFunc("basename", filepath.Base),
+	plugin.TemplateFunc("getService", getService),
+}
+
+const interfaceTemplate = `
+// Code generated by thriftrw --generate-plugin-api
+// @generated
+
+<$module := index .Request.Modules .Service.ModuleID>
+package <basename $module.Package>
+
+type <.Service.Name> interface {
+	<if .Service.ParentID>
```

Comment:
this is fine for now, but I imagine this pattern will exist in almost service generator. We should think about how to make this easier (even if it's just a function that will do all this that users will explicitly need to register using `TemplateFunc`)